### DISCUSSION TITLE
Using request queues with input.ffmpeg and smooth_add

### DISCUSSION THREAD

**Discussion by krt222** [savonet/liquidsoap#3807] [Q&A]
Hey, all!

I am playing around with a packaged version of liquidsoap (2.0.2) trying to understand the basics.
I would like to create a workflow that has one main feed (a sine) and a random list of one-off media that gets smooth_added to the main one. The URIs for the one-off media are written to a file and have to be played with input.ffmpeg (https links). The file of one-off media might be blank and is not always populated.

A barebones script that pretty much sums up my progress:

```
s = sine(100.0)
c = request.player()

oneoff_file = "/tmp/oneoff_media"
oneoff_content = {file.contents(oneoff_file)}
def getmedia()
  string.trim(file.contents(oneoff_file))
end
thread.when(getter.changes(oneoff_content), {c.play(request.create(getmedia()))})

o = smooth_add(normal = s, p = 0.30, special = c)
output(o)
```

If I write something to the oneoff_file (e.g. "https://..."), I get a log of ```Unknown protocol "https" in URI "https://...``` as well as ```Failed to prepare track: request not ready.```. The URLs are playable on their own with input.ffmpeg. I have not found a way to pass input.ffmpeg to c.

Is what I try to do feasible in liquidsoap in v.2.0.2 or in later versions? 
If so, are there any pointers as to how can I make request queues (play or push or dynamic) to get along with input.ffmpeg? I tried reading the book, but although it is well-written, I could not find something close to this example.

Thank you.

**Comment by vitoyucepi** [comment]
Hi @krt222,
Did you try to use `playlist(oneoff_file, reload_mode="watch")` instead of `request.player` and the following contents checker?

> If I write something to the oneoff_file (e.g. "https://..."), I get a log of Unknown protocol "https" in URI "https://... as well as Failed to prepare track: request not ready.. The URLs are playable on their own with input.ffmpeg. I have not found a way to pass input.ffmpeg to c.

I don't know about this, but there may be something wrong with the package. Are you running *ubuntu:jammy*?

**Comment by krt222** [comment]
> Hi @krt222, Did you try to use `playlist(oneoff_file, reload_mode="watch")` instead of `request.player` and the following contents checker?

I just tried it. Same error of ```Unknown protocol "https" in URI```. I would really like to get input.ffmpeg to play it since I would really like to experiment with containers that input.ffmpeg can process.


> I don't know about this, but there may be something wrong with the package. Are you running _ubuntu:jammy_?

Yes. If there is a way to do what I want to do in a later liquidsoap version, it's fine.

**Reply by vitoyucepi** [reply]
They probably did something wrong, anyway, here's the latest version.
https://github.com/savonet/liquidsoap/releases/tag/v2.2.4
Get a deb file for your operating system and install it.

You can also use the container image `savonet/liquidsoap:v2.2.4` or install from opam.
https://www.liquidsoap.info/doc-2.2.4/install.html

**Comment by krt222** [comment]
> They probably did something wrong, anyway, here's the latest version. https://github.com/savonet/liquidsoap/releases/tag/v2.2.4 Get a deb file for your operating system and install it.

Tried liquidsoap_2.2.4-ubuntu-jammy-2_amd64.deb, but now the log in both my initial script and the ```playlist...reload_mode="watch"``` method is:

 ```
[procol.external:3] Failed to find a file extension for "https://..." . Response has unknown mime-type: "audio/webm" you may want to add it to `settings.http.mime.extnames` and report to us if it is a common one.
[procol.external:3] Error while fetching http data: 522 - Connection timed out
[request.once:3] Could not resolve request https://...: failed.
```

Note that the 522 timeout is not a transient error. I can play the exact same media through input.ffmpeg, or even ffmpeg/ffplay at the command line, both before and after the failing test.

For example when using input.ffmpeg:
```
[decoder.ffmpeg:3] Requested content-type for "https://...": {audio=pcm(stereo)}
[decoder.ffmpeg:3] FFmpeg recognizes "https://...": {audio=pcm(stereo)}
```
...and the file plays as it should.

**Reply by vitoyucepi** [reply]
> Response has unknown mime-type

Just add it to the list of known mime types.
```ruby
settings.http.mime.extnames := ("audio/webm", ".webm")::settings.http.mime.extnames()
```
https://www.liquidsoap.info/doc-2.2.4/settings.html#mime-to-file-extension-mappings

**Reply by krt222** [reply]
This indeed removes the warning (thank you!), however the file still does not get played and a 522 error persists:

```
[procol.external:3] Error while fetching http data: 522 - Connection timed out
[request.once:3] Could not resolve request https://...: failed
```

**Reply by vitoyucepi** [reply]
Could you share url? Maybe site drops connections by useragent?
Or `curl -v https://...`.

**Reply by krt222** [reply]
I cannot share the exact URL unfortunately since it's on a private network, but thankfully I just found out that any URL from YouTube results in the exact same behavior. 

In order to do that just get an audio only stream URL with something like ```youtube-dl -f -g 250 https://www.youtube.com/watch?v=PyD9cMarVJk > /tmp/oneoff_media```  (This links to a sound of a sine). The resulting URL (...googlevideo.com/videoplayback?expire...) exhibits the same behavior ("Connection time out" etc when using either versions of my script, but correct playback if using input.ffmpeg on its own).

> Maybe site drops connections by useragent?

Neither YouTube (as far as I can tell) nor the internal server (I am 100% positive) drop connection by useragent.

**Reply by vitoyucepi** [reply]
Isn't *yt-dlp* a better version? Some time ago liquidsoap switched to it by default.
You could also use the *youtube-dl* protocol to get urls from youtube. Like `youtube-dl:uri`.

**Reply by vitoyucepi** [reply]
Of course it's a timeout. When I use curl to download 8M audio from youtube, it takes 3 minutes because the speed is 40k. After 30 seconds you'll see a timeout error.

**Reply by krt222** [reply]
Problem persists with yt-dlp. youtube-dl or yt-dlp does not really get a say in here since it is only used to get the actual URL for the playable media (the "...googlevideo.com/videoplayback..." thing).

**Reply by vitoyucepi** [reply]
https://github.com/savonet/liquidsoap/discussions/3807#discussioncomment-8878624

**Comment by Moonbase59** [comment]
Just gave the "playlist" variant a quick run-through and it works nicely:

```
# test_oneoff.liq

settings.log.level := 3
settings.http.mime.extnames := ("audio/webm", ".webm")::settings.http.mime.extnames()

s = sine(100.0)

c = playlist("test_oneoff.m3u", loop=false, reload_mode="watch")

o = smooth_add(normal=s, p=.3, special=c)

output.ao(o)
```

While running, I then give it something to work (_must_ be something playable!):

```bash
echo "https://radio.niteradio.net/api/station/1/ondemand/download/7d8da3f7dd71bb0bca343c2f" > test_oneoff.m3u
```

```bash
echo "https://radio.niteradio.net/api/station/1/ondemand/download/e13192f1ce6f8bc30c377e56" > test_oneoff.m3u
```

```bash
echo "youtube-dl:https://www.youtube.com/watch?v=zGDzdps75ns" > test_oneoff.m3u
```

```bash
echo "https://file-examples.com/storage/feaeb6f40265fd86e98cc0a/2017/11/file_example_MP3_700KB.mp3" > test_oneoff.m3u
```

```bash
echo "gtts:Hi there, I'm testing a one-off message." > test_oneoff.m3u
```

You can even _add_ stuff, like your YT video above:

```bash
echo "youtube-dl:https://www.youtube.com/watch?v=PyD9cMarVJk" >> test_oneoff.m3u
```

For reference: Used Liquidsoap 2.2.4-1 on Linux Mint 21.3.

**Comment by krt222** [comment]
Thanks for the answer. I can confirm that this works in they way you used iut, however it does not address the initial problem of the timeouts when using the final googlevideo address (youtube-dl/yt-dlp with the "-g" option). A small caveat (not completely applicable in my case) is that I didn't find a way to specify the format when using "youtube-dl:" : it downloads a 720p video too which might cause trouble to users that are on metered ISP connections.

So in order for me to not take more of your time, is there a way to force liquidsoap to use ffmpeg for the download of the stream? I mean, just like if you use:

```
output(input.ffmpeg("https://...googlevideo.com/videoplayback.."))
```

which plays the media correctly and does not result in timeouts, I would like to replicate this behavior (by using the googlevideo URL) in my script which uses request queues/player.


I tried to skim through the protocols documentation and experimented with "ffmpeg" and "process" protocols (i.e. "ffmpeg://https://...googlevideo.com/videoplayback...") but the first reported the same 522 error. The process protocol experiment failed too, but I am pretty positive this was partly because I did not understand how it works.

Thanks again for the time you people devote to this discussion.

**Reply by vitoyucepi** [reply]
> users that are on metered ISP connections.

Why not create a separate web application that allows users to download media from the internet or simply drag and drop their files?

**Comment by Moonbase59** [comment]
The argument of not having to download the whole video, but just the audio part instead is _valid_. It would indeed be nice having a `youtube-dl` protocol option or even a new protocol like `youtube-dl-audio` that could download _only_ the "best audio".

Problem with trying to find the (fast-expiring) direct Google link to _only the audio part_ of a video (like format 250 = Opus Audio) is that they stuff the link with lots of data, and try to restrict some things.

So getting the Opus Audio (only) from your video at [https://www.youtube.com/watch?v=PyD9cMarVJk](https://www.youtube.com/watch?v=PyD9cMarVJk) gives [https://rr3---sn-h0jeenle.googlevideo.com/videoplayback?expire=1711151696&ei=8MX9ZfPJMPCCi9oPyaSfuAM&ip=2003%3Ad6%3Aaf2f%3A4700%3A8eb4%3Aa842%3Ac025%3Aba4b&id=o-AEG3a09kpCPJWCg0wZxCKo7WkCyaTrqFl_Av1mv_3Dqd&itag=250&source=youtube&requiressl=yes&xpc=EgVo2aDSNQ%3D%3D&mh=l_&mm=31%2C26&mn=sn-h0jeenle%2Csn-4g5edns7&ms=au%2Conr&mv=m&mvi=3&pl=37&initcwndbps=1583750&spc=UWF9f0m0LbCk4RxmfvGHacxVnlNpPIljWrV0PwmzQYJP5YA&vprv=1&svpuc=1&mime=audio%2Fwebm&ns=tOzBCNF0JmkL0zkkOBvCDKYQ&gir=yes&clen=4807167&dur=365.521&lmt=1540780863664573&mt=1711129741&fvip=4&keepalive=yes&c=WEB&sefc=1&txp=5411222&n=w_3TApnIE5hODQ&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cxpc%2Cspc%2Cvprv%2Csvpuc%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AJfQdSswRQIhAMJaiwdBwQg-rNGYQUNyF-ISnnREczB35wr65H8V9IEVAiAPGVra_YgSBbfsUo8PSqDLEMElbsqixyYHApB-h6lCkA%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=ALClDIEwRgIhAN-QE029VwMZEfW-p0_hzxni5EVRygyCzwUDerRqrFl7AiEApXyAb_2eJ9pY6y_9demuOk0o6sP1vWPbRE7nzC4LUoo%3D](https://rr3---sn-h0jeenle.googlevideo.com/videoplayback?expire=1711151696&ei=8MX9ZfPJMPCCi9oPyaSfuAM&ip=2003%3Ad6%3Aaf2f%3A4700%3A8eb4%3Aa842%3Ac025%3Aba4b&id=o-AEG3a09kpCPJWCg0wZxCKo7WkCyaTrqFl_Av1mv_3Dqd&itag=250&source=youtube&requiressl=yes&xpc=EgVo2aDSNQ%3D%3D&mh=l_&mm=31%2C26&mn=sn-h0jeenle%2Csn-4g5edns7&ms=au%2Conr&mv=m&mvi=3&pl=37&initcwndbps=1583750&spc=UWF9f0m0LbCk4RxmfvGHacxVnlNpPIljWrV0PwmzQYJP5YA&vprv=1&svpuc=1&mime=audio%2Fwebm&ns=tOzBCNF0JmkL0zkkOBvCDKYQ&gir=yes&clen=4807167&dur=365.521&lmt=1540780863664573&mt=1711129741&fvip=4&keepalive=yes&c=WEB&sefc=1&txp=5411222&n=w_3TApnIE5hODQ&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cxpc%2Cspc%2Cvprv%2Csvpuc%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AJfQdSswRQIhAMJaiwdBwQg-rNGYQUNyF-ISnnREczB35wr65H8V9IEVAiAPGVra_YgSBbfsUo8PSqDLEMElbsqixyYHApB-h6lCkA%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=ALClDIEwRgIhAN-QE029VwMZEfW-p0_hzxni5EVRygyCzwUDerRqrFl7AiEApXyAb_2eJ9pY6y_9demuOk0o6sP1vWPbRE7nzC4LUoo%3D) _at this point in time, from my location_.

- Trying to download that with `curl` takes _minutes_ and is awfully slow.
- Using LS playlist with this link fails with a 522 HTTP error (CloudFlare/Google load balancer timeout).
- Same when using `ffmpeg:` plus the URI in the playlist.
- Using `ffmpeg` directly to download that is astonishingly fast, returning a file that is also _larger_ than the file downloaded using `curl`.

With all the data returned back to Google in the audio link, I very strongly suspect they simply don’t like something in our handling the download, be it only the user-agent or something.

**Conclusion:** Normal links and protocols work fine. Since we’re not here to circumvent Google trying to block downloading stuff from YouTube, but apparently `ffmpeg` can do it, you’ll most probably have to end up
- downloading the audio in your external process that _already_ can find the direct audio download link, and stuff _that_ filename into the playlist, or
- write your own protocol to handle that.

**Comment by krt222** [comment]
Thank you very much for the thorough answer.

> The argument of not having to download the whole video, but just the audio part instead is _valid_. It would indeed be nice having a `youtube-dl` protocol option or even a new protocol like `youtube-dl-audio` that could download _only_ the "best audio".

If someone happens to read this discussion in the future and would like to implement this, I would suggest to enable passing youtube-dl format options, with the possible default of "bestaudio".

> **Conclusion:** Normal links and protocols work fine. Since we’re not here to circumvent Google trying to block downloading stuff from YouTube, but apparently `ffmpeg` can do it, you’ll most probably have to end up
> 
> * downloading the audio in your external process that _already_ can find the direct audio download link, and stuff _that_ filename into the playlist, or

Downloading a file locally was never an option, since there would be an unavoidable delay for the completion of the transfer. 

> * write your own protocol to handle that.

Yeah, I think that's the way to go.


Thank you again for the useful information. Much appreciated.

**Comment by Moonbase59** [comment]
I just tried a quick-n-dirty hack, copied the `youtube-dl`  protocol as `youtube-dl-audio` (using `-f bestaudio`) and it indeed seems to work:

```
# test_oneoff.liq

settings.log.level := 3
settings.http.mime.extnames := ("audio/webm", ".webm")::settings.http.mime.extnames()

# Register the youtube-dl-audio protocol, using youtube-dl.
# Syntax: youtube-dl-audio:<ID>
# @flag hidden
def protocol.youtube_dl_audio(~rlog, ~maxtime, arg) =
  binary = settings.protocol.youtube_dl.path()
  timeout = settings.protocol.youtube_dl.timeout()

  def log(~level, s) =
    rlog(s)
    log(label="protocol.youtube-dl-audio", level=level, s)
  end

  delay = maxtime - time()
  cmd =
    "#{binary} --get-title --get-filename -- #{process.quote(arg)}"
  log(
    level=4,
    "Executing #{cmd}"
  )
  x = process.read.lines(timeout=delay, cmd)
  x = if list.length(x) >= 2 then x else ["", ".osb"] end
  title = list.hd(default="", x)
  ext = file.extension(leading_dot=false, list.nth(default="", x, 1))
  cmd =
    "#{binary} -q -f bestaudio --no-continue --no-playlist -o $(output) -- #{process.quote(arg)}"

  cmd = process.uri(timeout=timeout, extname=ext, cmd)
  if
    title != ""
  then
    ["annotate:title=#{string.quote(title)}:#{cmd}"]
  else
    [cmd]
  end
end

protocol.add(
  "youtube-dl-audio",
  protocol.youtube_dl_audio,
  doc=
    "Resolve a request using youtube-dl. This downloads audio only.",
  syntax="youtube-dl-audio:uri"
)



s = sine(100.0)

c = playlist("test_oneoff.m3u", loop=false, reload_mode="watch")

o = smooth_add(normal=s, p=.3, special=c)

output.ao(o)
```

Run it and add the "normal" YT link to the playlist as:

```
echo "youtube-dl-audio:https://www.youtube.com/watch?v=PyD9cMarVJk" > test_oneoff.m3u
```

@toots: Could this (using `-f bestaudio` instead of `-f best`) be somehow added as an option to the existing `youtube-dl` protocol? Or do we really need an extra protocol?
